Pattern overrides: use block binding editing API

ellatrix opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Updates the pattern (`wp:block`) block to use the block bindings API (from #60724) in the editor.

## Why?
This fixes a specific use case where two blocks (e.g. two paragraphs) may use the same names for the pattern overrides. In this case the two blocks should be synchronized when editing, but this is broken in `trunk`.

## How?
Add new pattern overrides binding source properties for getValue and setValues

## Testing Instructions
1. Create a synced pattern with two paragraphs
2. Enable overrides for the two paragraphs and use the same block names for both
3. Save the pattern
4. Create a new post and insert the pattern
5. Try typing in one of the paragraphs

Expected: The two paragraphs can be synchronized.

## Screenshots or screencast <!-- if applicable -->

kevin940726 commented:
I think this PR is intended to address https://github.com/WordPress/gutenberg/issues/59817? Is the work @SantosGuillamot mentioned in the issue? There seem to be some failing e2e tests. Is this ready for review or still an experiment/WIP?

ellatrix commented:
This is ready for review, I just need to adjust the e2e tests. For some reason those blocks were added as inner blocks where there are actually controlled blocks from an entity? Not sure I understand why those tests are failing.

kevin940726 commented:
I briefly tried it and noticed this thing:

https://github.com/WordPress/gutenberg/assets/7753001/08956e8c-367b-43f5-bee6-adbab86396f3

The original content is lost as soon as the block allows overrides. I'm not sure if it's the expected behavior for other sources, but for pattern overrides, we want to leave the original content intact. Would it be something that needs to be special cased?

talldan commented:
> The original content is lost as soon as the block allows overrides.

Seeing a similar thing here, though additionally a few further issues with editing the original pattern. Any existing blocks in a saved pattern seem to be uneditable (e.g. create a pattern, save it, reload, try editing the original pattern content). Blocks with bindings can be edited, but the edits aren't saveable.

It might be more of an issue with the upstream branch.

talldan commented on packages/editor/src/bindings/pattern-overrides.js:
```diff
@@ -2,10 +2,53 @@
+		const patternClientId = parents.find(
+			( id ) => getBlockName( id ) === 'core/block'
+		);
+		const blockName = currentBlockAttributes?.metadata?.name;
```
Nitpick - this line could be moved to after the early return.

SantosGuillamot commented:
I've rebased the branch and added a these commits to solve a couple of issues I found:
* Run only the attributes included in `nextAttributes`: [link](https://github.com/WordPress/gutenberg/pull/60721/commits/6cd3d3bd3920b990f0614aa933161b4ec9b53803). Prior to that, we were iterating over all the attributes included in the bindings object, but I think we want to iterate over the ones being updated in `nextAttributes.
* Update the `getValue` and `setValue` functions when editing the pattern: [link](https://github.com/WordPress/gutenberg/pull/60721/commits/90535fb452a71d01382f03ba7dd5c14a7260d63e). If I am not mistaken, when editing a pattern, even if they have bindings, we want to keep the normal behavior of showing the default value and updating the attribute itself instead of the pattern block.

Apart from that, there are other issues we need to work on:

* Issue when replacing image

Right now, when you replace an image, it doesn't only update the `url` attribute, but a bunch of them: [link](https://github.com/WordPress/gutenberg/blob/1f6fec6575ca4e9bcd64822b533df9f3057b9ba1/packages/block-library/src/image/edit.js#L255-L259). It includes id, link, href, sizeSlug, linkDestination. As those attributes are not supported yet by the block bindings, they don't work as expected. I believe that just supporting them should solve the issue. And, if I am not mistaken, it should be a problem with the current implementation as well.

* Issue when changing button URL

The same issue we have with the image happens in the button. When the user enters a URL, it updates more attributes: [link](https://github.com/WordPress/gutenberg/blob/1f6fec6575ca4e9bcd64822b533df9f3057b9ba1/packages/block-library/src/button/get-updated-link-attributes.js#L50-L53). In this case, `rel` is not supported yet and it doesn't work as expected. Again, it could potentially be solved by supporting it.

* Not showing the limited block controls toolbar

I can see that with this refactoring, we are showing all the block controls, while in the existing implementation it is limited. We probably removed some code handling that logic we might need to revisit.

kevin940726 commented:
Thanks for following up on this! I also noticed that the named block that doesn't have a binding is no longer editable even in the pattern editor.

https://github.com/WordPress/gutenberg/assets/7753001/f5c9054c-e640-486c-803b-e77b70e71d0d

I'm not sure if it's an issue from my end though. Would appreciate a sanity check!

SantosGuillamot commented:
I've been triaging it a bit more and it seems it works when I go to the code editor and modify anything:


https://github.com/WordPress/gutenberg/assets/34552881/037089bd-42ac-407d-a404-fca6ec839406

It seems that, when we are in the page, we set the edit mode [here](https://github.com/WordPress/gutenberg/blob/90535fb452a71d01382f03ba7dd5c14a7260d63e/packages/block-library/src/block/edit.js#L85-L88):
* For blocks with bindings we set it to `contentOnly`.
* For blocks without bindings we set it to `disabled`.

So far everything seems fine.

However, I assume we should reevaluate that after navigating to the "Edit Original" through `onNavigateToEntityRecord`: [link](https://github.com/WordPress/gutenberg/blob/90535fb452a71d01382f03ba7dd5c14a7260d63e/packages/block-library/src/block/edit.js#L220-L223).

Any idea how it can be solved? I couldn't figure out why it is working on `trunk` but not in this PR. Maybe a missing dependency?

youknowriad commented:
That's just another bug that results of the multiple calls we have to `setBlockEditingMode` action. Unfortunately, we made a mistake when we introduced this API. It is an API that results very quickly in race conditions (like seen here but also seen elsewhere, zoom-out mode was one)

Basically there's probably two components "competing" to set the right editing mode for blocks: one might be the blocking bindings content locking thing, and the other could be the template-lock rendering mode that enables/disables blocks based on whether they are post blocks or not and maybe something else as well.

---

There might be a hacky fix or temporary fix here but the "right fix" is to actually remove any usage of `setBlockEditingMode` from our code base and instead have the `getBlockEditingMode` figure out the right editing mode more declaratively (based on modes, on containers, ...)

youknowriad commented on packages/block-library/src/block/edit.js:
```diff
@@ -314,78 +178,25 @@ function ReusableBlockEdit( {
-
-				replaceInnerBlocks( patternClientId, blocks );
-			} );
-			setBlockEditingMode( patternClientId, originalEditingMode );
```
The fact that we were making these calls here in trunk could have "hidden" the race conditions. It doesn't mean the race conditions didn't exist on trunk, they did, but they were hidden or maybe they resulted in different bugs (like things being editable in the template while in "template-lock" mode)

SantosGuillamot commented:
Thanks a lot for the context, Riad. It's really useful to understand the situation.

I've been taking a deeper look, and I found a potential fix/issue reflected in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/7958101b90b34b728b5a5adbcbc9fe776439f39d). Let me try to explain it:

* In this PR, we are relying on `blocks` obtained from `useEntityBlockRecord`: [link](https://github.com/WordPress/gutenberg/pull/60721/files#diff-3e104de22b2ea1fac4d9f2fd42f07a1c8ff39bcb9d0bfa43b056538e74e1c558R140-R144).
* Previously, we were relying on `getBlocks( patternClientId )`: [link](https://github.com/WordPress/gutenberg/pull/60721/files#diff-3e104de22b2ea1fac4d9f2fd42f07a1c8ff39bcb9d0bfa43b056538e74e1c558L294).
* However, it seems that blocks obtained from `useEntityBlockRecord` don't have the correct clientId, while the ones obtained from `getBlocks` do.
* That means that if we call `setBlockEditMode` with the incorrect block ids, it doesn't work as expected: [link](https://github.com/WordPress/gutenberg/blob/7958101b90b34b728b5a5adbcbc9fe776439f39d/packages/block-library/src/block/edit.js#L186). (I changed it in the latest commit, previously it was using blocks instead of innerBlocks.

So, it seems to fix the issue, but I'm not sure why `blocks` obtained from `useEntityBlockRecord` don't have the correct ids. Is this expected?

I've made a video trying to explain myself:


https://github.com/WordPress/gutenberg/assets/34552881/11fcde82-e9b0-405f-91be-18da04777e95

youknowriad commented:
> So, it seems to fix the issue, but I'm not sure why blocks obtained from useEntityBlockRecord don't have the correct ids. Is this expected? It seems they are helpful [here](https://github.com/WordPress/gutenberg/pull/60721/files#diff-3e104de22b2ea1fac4d9f2fd42f07a1c8ff39bcb9d0bfa43b056538e74e1c558R215-R221) because using innerBlocks breaks it.

If I'm not wrong `useBlockSync` clone the blocks internally to be able to have the same reusable blocks in multiple places at the same time which means the clientIds outside and inside (on initial render only) will differ. I think @ellatrix was trying to remove that cloning but it might be challenging.

ellatrix commented:
Ah, the cloning. Yes, tried to remove it in #54737, but I didn't really try to find a solution. It was first introduced in #27885.

ellatrix commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -115,25 +115,28 @@ export const withBlockBindingSupport = createHigherOrderComponent(
-					for ( const [
-						attributeName,
-						boundAttribute,
-					] of Object.entries( bindings ) ) {
```
Could you leave an inline comment on why you loop over the next attributes instead of bindings?

SantosGuillamot replied:
Added in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/1aa942e58054f713acd8bdcd2e6224ff8479119d). Let me know if that's not enough 🙂

ellatrix commented on packages/block-library/src/block/edit.js:
```diff
@@ -444,7 +230,7 @@ function ReusableBlockEdit( {
 	const resetContent = () => {
 		if ( content ) {
-			replaceInnerBlocks( patternClientId, initialBlocks );
+			setAttributes( { content: undefined } );
```
What is the content attribute?

SantosGuillamot replied:
In the pattern block, when you override a block attribute, it adds a new "content" attribute pointing to the name of the block. Something like this:

`<!-- wp:block {"ref":1602,"content":{"Custom Name":{"content":"Edited paragraph content"}}} /-->`

ellatrix replied:
I wonder if there's a better attribute name we can use. Maybe `overrides`?

ellatrix replied:
Also, do we declare this attribute in the block type registration?

ellatrix replied:
Ah, I see this is a preexisting attribute

kevin940726 replied:
> I wonder if there's a better attribute name we can use. Maybe `overrides`?

It's used to be called `overrides`, but the feedback was to change it to `content`.

talldan commented:
This seems to be working much more reliably now. 👏  

One bug I can see is that the 'Reset' button on individual blocks doesn't seem to work:
![Screenshot 2024-04-26 at 2 54 07 pm](https://github.com/WordPress/gutenberg/assets/677833/c8ca754f-24bf-4f32-aebf-ff444447bc81)

talldan commented on packages/patterns/src/components/reset-overrides-control.js:
```diff
@@ -6,80 +6,52 @@ import {
-	const patternWithOverrides = useSelect(
+	const { updateBlockAttributes } = useDispatch( blockEditorStore );
+
+	const { isOverwritten, resetOverrides } = useSelect(
```
`isOverridden` would be a better name IMO.

SantosGuillamot replied:
Totally right. Changed in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/501ade3c978108dccc87f4e6b94124772351b6b7).

talldan commented:
Thanks for the quick fix! It's looking good. I guess we need to figure the best way to update the failing e2e tests. I haven't looked too closely into it, but maybe @kevin940726 has some ideas.

Updating the PR description would be great as well.

The only slightly unusual thing I see in this PR compared to trunk is that undo/redo is a little glitchier. Almost like the pattern content is momentarily completely removed and added back again. It does work though, so it could be a follow up to investigate this. I don't know if the same happens for other implementations of the block bindings API.

SantosGuillamot commented:
> Thanks for the quick fix! It's looking good now.

I was about to share that I think with this new approach, we can simplify the logic following something similar to the [pattern-overrides bindings file](https://github.com/WordPress/gutenberg/pull/60721/files#diff-deffea7af4722e1274cca18247b89c007e041054cc2fa315c68a0bb1425d9eb8R36-R66). If I am not mistaken, in this case we just want to modify the pattern block.

I pushed [the commit](https://github.com/WordPress/gutenberg/pull/60721/commits/615148b31f954b6ae0086f33df79aead9232bb12) with a potential path forward. However, I am not familiar with that part of the code, so I may be missing something.

> I guess we need to figure the best way to update the failing e2e tests. I haven't looked too closely into it, but maybe @kevin940726 has some ideas.

I'll rebase this branch to resolve the conflicts and check if the tests are still failing and why.

> Almost like the pattern content is momentarily completely removed and added back again.

Could you share the steps to reproduce? I can try to take a look at it as well.

SantosGuillamot commented:
I have one question that might affect this PR: What should happen when two blocks have the same name/id? Should they be synced both when editing a pattern in a post and when editing the pattern itself?

If that's the case, I can make a commit to make it work that way.

I made a quick video to illustrate what I mean:


https://github.com/WordPress/gutenberg/assets/34552881/6546f7ca-3d16-4a71-ab24-be5ecc8793d5

gziolo commented:
@SantosGuillamot, the demo included two blocks sharing the same name. Looks great. If they share the block type then syncing them seems like an interesting feature. If they are different block types, then it becomes difficult because they might use different attribute names, so I don't know what the system would do in that case.

youknowriad commented:
> If they are different block types, then it becomes difficult because they might use different attribute names, so I don't know what the system would do in that case.

It would be cool to allow different block types to use the same value if they share the same name. For instance I can add a "company name" bound property and use in a both a heading and a paragraph or something.

SantosGuillamot commented:
I've implemented an initial version of the syncing in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/3611bae3c6b54a58dcf579f9ba288c390d0ff2cf). I'm assuming that if the blocks share the same name and the same attribute name they can by synced. We can iterate on that (or remove it).

SantosGuillamot commented:
I've also been looking at the synced patterns (no overrides) tests and I made a first attempt to "fix" them in [this other commit](https://github.com/WordPress/gutenberg/pull/60721/commits/a9c7e3e69d3ed399a393c241dca2462671a9492a). The idea is that we don't check only the editor but we check that we are actually reading/editing the entity record.

I have some aspects to review, though:

* I had to fix the detach functionality in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/99b43d0ad4db082b10084cac2c761a6ee8df2e6e).
* In some tests, we have to wait until the pattern is inserted or created. I wasn't sure how to do that.
* When we edit the pattern but don't save it, it seems that the changes are stored in a new `blocks` property while `content` is undefined until it is saved. Is this supposed to work that way?

I'll take a look at the remaining failing pattern overrides tests.

kevin940726 commented on test/e2e/specs/editor/various/patterns.spec.js:
```diff
@@ -149,28 +157,22 @@ test.describe( 'Synced pattern', () => {
-				},
-			] );
-		const after = await editor.getBlocks();
+		// Wait until the pattern is inserted.
+		await editor.canvas
+			.getByRole( 'document', {
+				name: 'Block: Pattern',
+			} )
+			.waitFor();
+
+		const [ syncedPattern ] = await editor.getBlocks();
+		const patternRecord = await getPatternRecord(
+			page,
+			syncedPattern?.attributes?.ref
+		);
+
+		expect( patternRecord?.content ).toBe(
+			'<!-- wp:paragraph -->\n<p>A useful paragraph to reuse</p>\n<!-- /wp:paragraph -->'
+		);
```
I don't think this is necessary. We can just assert that the post content includes the `<!-- wp:block -->` block. Then we can visit the frontend to verify that the rendered post includes the pattern content (paragraph in this case.)

This kind of test only has to be done once and we only have to check that it contains `wp:block` in the remaining tests.

I think this also matches more closely to what the end users would expect. The tests are only what they seem because it's easier this way in the previous implementation, but they are technically testing implementation details.

talldan commented:
> It would be cool to allow different block types to use the same value if they share the same name. For instance I can add a "company name" bound property and use in a both a heading and a paragraph or something.

It worked well when I tested it. It might be a good case for e2e tests two.  Thinking on it though, there's at least one odd case, image and buttons both have `url`, but they mean different things (one is the `src`, one is the link `href`). 

The button block could be updated to use `href` as its attribute name for consistency, and then it would work as expected. Changing a block's attribute name has an unfortunate side-effect, anyone using a button in a pattern override up till now will have their content disappear because we lose the binding. 

For other types of block bindings, a deprecation can be added, but for pattern overrides, migrating data in the `content` object of the pattern block is trickier, it might have to be a back compat break for the plugin.

> Could you share the steps to reproduce? I can try to take a look at it as well.

@SantosGuillamot Sure, though I'd definitely not consider it a blocker for merging this PR. I'd be pretty keen to get this merged earlier, as some other tasks are partly dependent on it.

1. Create a synced pattern with overrides, use a few different block types in the pattern
2. Insert an instance of the pattern somewhere and add content to the overridden blocks
3. Use the 'Reset' button on the pattern block
4. Now use undo and redo, and notice there seems to be very brief a flash of missing content before the correct content is shown

I'm not sure that it's related to 'Reset', it's a bit easier to reproduce using that button though as it changes the entire pattern at once.

SantosGuillamot commented:
I was looking at the pull request and, apart from the shared comments, there are a couple of existing issues that we need to solve before this PR gets merged:

## 1. Calling `setAttributes` for non-connected block attributes

There is a broken experience when we are overriding attributes of a pattern in a post/page and we call `setAttributes` for attributes that are not connected. This happens when you replace an image, for example. The "Replace" button doesn't only update the url attribute, but a bunch of them: [link](https://github.com/WordPress/gutenberg/blob/1f6fec6575ca4e9bcd64822b533df9f3057b9ba1/packages/block-library/src/image/edit.js#L255-L259). It includes id, link, href, sizeSlug, linkDestination.

I've recorded this video to show what I'm referring to:

https://github.com/WordPress/gutenberg/assets/34552881/d480ee5e-80ee-46ef-a037-59e999d3fc05

I've started a [branch](https://github.com/WordPress/gutenberg/compare/trunk...add/support-all-image-attributes-in-block-bindings#) (and I'll open a PR) to support all image block attributes. That should partially solve the issue.

However, if, at any point, it is possible only to override some attributes of a block, this would still be a problem.

The only thing I can think of is to filter somehow the `setAttributes` when we are overriding a pattern only to include the ones that are bound. Something like this (the code used in the video):

```js
if ( patternClientId ) {
  for ( const attributeName in nextAttributes ) {
    if ( ! bindings[ attributeName ] ) {
      delete keptAttributes[ attributeName ];
    }
  }
}
```

However, if we do that, I am not sure where it should go.

I am not sure why this isn't a problem with the current implementation in `trunk`.

Any ideas on how this could be approached?

## 2. "resets overrides after clicking the reset button" test failing

It seems that test is failing because the "Undo" button is not reliable. I've tested it locally and I haven't been able to reproduce it though, it seems to work fine for me.

I assume it could be related to the issue shared by @talldan [here](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2084313263).

I'll try to take another look, but any idea on why this could be happening, or if you are able to reproduce, is highly appreciated.

ellatrix commented:
> This happens when you replace an image, for example. The "Replace" button doesn't only update the url attribute, but a bunch of them: [link](https://github.com/WordPress/gutenberg/blob/1f6fec6575ca4e9bcd64822b533df9f3057b9ba1/packages/block-library/src/image/edit.js#L255-L259). It includes id, link, href, sizeSlug, linkDestination.

All these attributes are connected, and that's something we discussed as well when trying to isolated controls by attribute key. This is where things get messy.

talldan commented:
> I am not sure why this isn't a problem with the current implementation in trunk.

The implementation in `trunk` doesn't use `useEntityBlockEditor` so there can never be any edits to the original pattern, instead it creates a copy of the blocks and edits are made to those.

I think it could be possible to filter out the edits to non-bound attributes by wrapping `onChange`/`onInput` in the pattern block and ensuring we don't call through to the original functions for non-bound attributes. It might not be a long term solution. 🤔

SantosGuillamot commented:
I've added a [workaround](https://github.com/WordPress/gutenberg/pull/60721/commits/51fe5e3bef219a023bb4a0187ad39393c13f4095) to solve the issue with `setAttributes` until we figure out if it should be solved by block bindings or it is specific to pattern overrides. It could be the case where this is needed by other sources as well.

I'm taking a look now at another issue. It seems that the Save button is not enabled after changing the "content" attribute of the pattern block when an image is inserted. I made a quick video to show what I mean:


https://github.com/WordPress/gutenberg/assets/34552881/324e0feb-3e0d-430f-8481-215e27eb9a36

SantosGuillamot commented:
So, as shown in [this video](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2097706514), it seems that the "content" block attribute is updated correctly when replacing the image, as shown in the Code Editor. However, it doesn't enable the "Save" button.

We are updating that attribute using `registry.dispatch( blockEditorStore ).updateBlockAttributes()` : [link](https://github.com/WordPress/gutenberg/pull/60721/files#diff-deffea7af4722e1274cca18247b89c007e041054cc2fa315c68a0bb1425d9eb8R66-R76).

Shouldn't that be enough? Any clue why it is not working as expected or where to start triaging it? @youknowriad @ellatrix

youknowriad commented:
It should be enough. We should check whether the `onChange` of `useBlockSync` is called or not. If there's no undo, I suspect that it's not called but I don't know why.

youknowriad commented:
I think onInput and onChange shouldn't have an impact on the "save" button. Both mark the post as dirty, the only difference is on undo/redo, so maybe there's something else going on here.

SantosGuillamot commented:
It seems the main difference between onInput and onChange is this content property: https://github.com/WordPress/gutenberg/blob/fd2e5ddf8c7a809223376c02f3c52feff48c847f/packages/core-data/src/entity-provider.js#L231-L232

If I add that to the onInput function, it seems to work as expected. But I don't fully understand that part yet, so I am not sure if it should be there or not.

SantosGuillamot commented:
Perfect 🙂 I've opened a [draft pull request](https://github.com/WordPress/gutenberg/pull/61496), copying and pasting the solution for `onChange`, to see if there are any issues with the tests. I'll keep exploring that.

SantosGuillamot commented:
Oh, I've just seen that it enables the "Save" button but there's no undo yet.

SantosGuillamot commented:
Taken a look at [the onInput and onChange definition](https://github.com/WordPress/gutenberg/blob/2dd15ff2432ec7ad0099aaa5b1594702e12f814c/packages/core-data/src/entity-provider.js#L142-L146) I'm wondering if we should ensure that any changes made to the `wp-block` attributes are persistent and they go through `onChange`, because we actually want to create an undo level, right?

> `onInput` is for block changes that don't create undo levels
> or dirty the post, non-persistent changes, and `onChange` is for
> persistent changes.

Maybe it still makes sense to add `content` to `onInput`, but I am not sure if that would solve the issue here if I am understanding it correctly.

SantosGuillamot commented:
Not sure if it is the right approach, but it seems that calling `__unstableMarkLastChangeAsPersistent` after [updating the block attributes](https://github.com/WordPress/gutenberg/blob/fd2e5ddf8c7a809223376c02f3c52feff48c847f/packages/editor/src/bindings/pattern-overrides.js#L66-L76) solves the issue.

```
registry
    .dispatch( blockEditorStore )
    .updateBlockAttributes( patternClientId, {
	[ CONTENT ]: {
		...currentBindingValue,
		[ blockName ]: {
			...currentBindingValue?.[ blockName ],
			[ attributeName ]: value,
		},
	},
} );

__unstableMarkLastChangeAsPersistent();
```

SantosGuillamot committed:
Update only the needed attributes with bindings

SantosGuillamot committed:
Use inner blocks for set edit mode

SantosGuillamot committed:
Keep blocks as normal when editing pattern

ellatrix committed:
Pattern overrides: use block binding editing API

SantosGuillamot committed:
Move variable after early return

SantosGuillamot committed:
Add inline comment for bindings setAttributes

SantosGuillamot committed:
Change indiviual reset logic

SantosGuillamot committed:
Fix detaching of synced patterns

SantosGuillamot committed:
Sync values of bound attributes

SantosGuillamot committed:
Change variable to isOverriden

SantosGuillamot committed:
Use overriden values when detaching pattern

SantosGuillamot committed:
Adapt synced pattern tests

SantosGuillamot committed:
Adapt pattern-overrides tests

SantosGuillamot committed:
Change missing isOverriden

SantosGuillamot committed:
Ensure change is marked as persistent in pattern overrides

SantosGuillamot committed:
Add workaround for pattern overrides

SantosGuillamot committed:
Stop using syncDerivedUpdates

SantosGuillamot committed:
Remove `content` pattern property if empty

SantosGuillamot committed:
Try to solve undo/redo issues

SantosGuillamot committed:
Add test for blocks with a shared name

SantosGuillamot committed:
Solve undo/redo issues

SantosGuillamot committed:
Manage undefined attrs in pattern overrides

SantosGuillamot committed:
Support button rel in bindings editor logic

SantosGuillamot committed:
Don't update non-bound attributes when using patterns

SantosGuillamot commented:
Okay, I've been working on it and it seems everything is "working" and tests are passing. BUT, there are a few things I am not sure about.

### 1. `blocks` from `useEntityBlockEditor` not using the proper clientIds cause undo/redo issues

As explained [here](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2077808890), the clientIds outside and inside (on initial render only) will differ, and that causes issues with undo and redo.

I "fixed" it in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/a95749f899d4e9cdff401fda7bc6c79e00d6819f) using the `innerBlocks` obtained using `getBlocks` when they are available, but I am not sure if it is the right approach. I couldn't use the `innerBlocks` directly because they are not available since the beginning.

I assume another possibility would be to replace the `blocks` clientIds from the ones we obtain from `innerBlocks`.

Any idea how this could be properly solved?

### 2. Mark changes as persistent after updating pattern block attributes

In order to solve [the issue with the "Save" button](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2097706514) (check the following comments for more context), I finally used in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/1101667f32946fcdfb0855a5fe7b048beb97b8c7) `__unstableMarkLastChangeAsPersistent` to ensure the changes are marked as persistent in pattern overrides.

It seems to be working. However, I don't know if that's correct or why it is even needed.

### 3. Workaround to not update non-bound block attributes

To solve this issue described [here](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2084940845), I added [this workaround](https://github.com/WordPress/gutenberg/pull/60721/files#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R119-R142) to the block bindings editor hook. Right now, it is specific to pattern, so it seems it shouldn't belong there, but it's true that we still have to understand better if that should be the case for the rest of the sources or how it should work. Additionally, I couldn't find another way of dealing with that.

Is this okay or should we keep looking for alternatives?

### 4. Should we remove `syncDerivedUpdates` action?

After this refactoring, it seems that [`syncDerivedUpdates` action](https://github.com/WordPress/gutenberg/blob/9d444a337ec478f4fdf3f6b3646abc55bd71388f/packages/block-editor/src/store/private-actions.js#L302) is not used anywhere. Could this cause any issues? Should we remove that action?

### 5. Changes to block bindings editor hook loop

While this pull request, we realized that there were [some changes needed](https://github.com/WordPress/gutenberg/pull/60721/files#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R144-R169) for the block bindings editor hook. I should have created a different pull request for that. Is it okay to keep it here or is it still better to move that code to a different PR?

youknowriad commented:
> __unstableMarkLastChangeAsPersistent

If you do that, you're forcing each change to bount attributes to create an undo level. But we don't want that every time, for instance when typing in a RichText we group some characters together, or when you edit the same attribute multiple times in a row, the editor has a built-in behavior that only creates a single undo level.

youknowriad commented:
Did you try wrapping the `source.setValue` calls and the remaining `setAttributes` calls within a single registry.batch?

SantosGuillamot commented:
The existing hook is wrapping the whole new `_setAttributes` within a registry.batch: [link](https://github.com/WordPress/gutenberg/blob/9d444a337ec478f4fdf3f6b3646abc55bd71388f/packages/block-editor/src/hooks/use-bindings-attributes.js#L112-L174). Is that what you mean?

youknowriad commented:
Mmm I missed it, Did you try without it and see what impact it has?

SantosGuillamot commented:
> Did you try without it and see what impact it has?

It seems to have the same behavior.

talldan commented:
> 2. Mark changes as persistent after updating pattern block attributes
> In order to solve https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2097706514 (check the following comments for more context), I finally used in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/1101667f32946fcdfb0855a5fe7b048beb97b8c7) __unstableMarkLastChangeAsPersistent to ensure the changes are marked as persistent in pattern overrides.

> It seems to be working. However, I don't know if that's correct or why it is even needed.

Do we know why the image block works differently when in a pattern vs an image block that's inserted into normal post content?

It's difficult to know what the right fix is without knowing the root cause.

> 3. Workaround to not update non-bound block attributes
> To solve this issue described https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2084940845, I added [this workaround](https://github.com/WordPress/gutenberg/pull/60721/files#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R119-R142) to the block bindings editor hook. Right now, it is specific to pattern, so it seems it shouldn't belong there, but it's true that we still have to understand better if that should be the case for the rest of the sources or how it should work. Additionally, I couldn't find another way of dealing with that.
> Is this okay or should we keep looking for alternatives?

This feels more general than pattern overrides. I don't think there should be any need to call `setValue` for a non-bound attribute. My feeling is that it should fallback to the normal `setAttributes` behavior in that situation. So when a block calls `setAttributes` with a mixture of bound and non-bound attributes, I think it should be divided into two sets, one that's passed to `setValue` for bound attributes, and one that's passed to the original `setAttributes` function for non-bound.

The problem for patterns is that we also need to avoid updating the `wp_block` entity when those non-bound attributes are updated, but I think there might be other ways to do that (potentially we can avoid passing `onChange` and `onInput` into `useInnerBlocksProps`.

> 4. Should we remove syncDerivedUpdates action?
> After this refactoring, it seems that [syncDerivedUpdates action](https://github.com/WordPress/gutenberg/blob/9d444a337ec478f4fdf3f6b3646abc55bd71388f/packages/block-editor/src/store/private-actions.js#L302) is not used anywhere. Could this cause any issues? Should we remove that action?

Probably best to do this in a separate PR (cc @kevin940726).

SantosGuillamot commented:
> Do we know why the image block works differently when in a pattern vs an image block that's inserted into normal post content?

The farthest I have been able to dig is [this](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2103020961):
* In a normal image we call `setAttributes` with a bunch of attributes, and the whole change is marked as persistent.
* In bindings, we call `updateBlockAttributes` for each attribute individually. The first change is marked as persistent, but the others are marked as not persistent. This means that when we finish the loop, the last change is not persistent when there is more than one attribute bound.

We could check `isLastBlockChangePersistent()` after each `setValue` and, if any of those is true, call `__unstableMarkLastChangeAsPersistent()` when the loop finishes. But I am not sure if that's correct or if it is even safe to assume that's always the case.

>  I don't think there should be any need to call setValue for a non-bound attribute. My feeling is that it should fallback to the normal setAttributes behavior in that situation.

This is how it is working right now. We only call setValue for bound attributes and we still call setAttributes for non-bound attributes. The problem is that for pattern overrides we don't want to call setAttributes at all for non-bound attributes.

> there might be other ways to do that (potentially we can avoid passing onChange and onInput into useInnerBlocksProps in the block edit function)

I will explore that possibility.

ellatrix commented on packages/patterns/src/components/reset-overrides-control.js:
```diff
@@ -6,80 +6,53 @@ import {
+
+			return {
+				isOverriden: !! existingOverrides?.[ name ],
+				resetOverrides: async () => {
```
Why is this async?

SantosGuillamot replied:
I'd say it is a copying and pasting issue from [here](https://github.com/WordPress/gutenberg/pull/60721/files/9d444a337ec478f4fdf3f6b3646abc55bd71388f#diff-eb4c36af06624e9031e5e7ed7850f9c13e5efe755f84ac044f424453571e2254L51-L76). I don't think it is needed.

ellatrix commented:
> > Did you try without it and see what impact it has?
> 
> It seems to have the same behavior.

I tried it without batch and it seems to fix the issue (after I also removed the `__unstableMarkLastChangeAsPersistent` fix). So why is `batch` breaking it? 🤔

talldan committed:
Remove pattern-overrides specific code from `withBlockBindingSupport`, instead NOOP onChange and onInput

Revert "Don't update non-bound attributes when using patterns"

This reverts commit ef29ab3caec7f81ec0d16f135f71b591b63710d3.

NOOP onChange and onInput in the pattern block - the original entity should never be updated via this code

talldan commented:
@SantosGuillamot I've pushed e28da536aac278d2f372019a49c5c0919d2d8c82 to demonstrate what I mean.

The block bindings code will try calling `setAttributes` for non-bound attributes in pattern instances, but because `onChange` and `onInput` are no longer used in the pattern block, there can never be any edits to the original pattern from a pattern instance. I think this is a good idea while the UX works as it does—the user can't edit the original pattern in any way from an instance. If it changes in the future it can be revised.

ellatrix commented:
These are the previous and next actions that run:

<img width="1352" alt="image" src="https://github.com/WordPress/gutenberg/assets/4710635/82c6bfeb-fb24-4d8f-88ae-fe5c9bbeb3c8">

`isPersistent` is controlled by `isUpdatingSameBlockAttribute`. In the case of block bindings with pattern overrides, we update the same attribute multiple times, anything but the first update will make `isPersistent` false. It all still works fine without batching, because all store listeners were fired for the first action, making things dirty. But with batching, `isPersistent` was never true at any time.

So there's two solutions I think:

* Either we change the API to allow updating a binding from multiple sources.
* Or we could potentially change `isUpdatingSameBlockAttribute` to ignore complex attributes (that are a plain object).

The first is probably the best. Can look into that in a bit.

ellatrix committed:
Fix persistent issues with setValues API

ellatrix commented:
I fixed it by adding a `setValues` API, which is cleaner anyway, instead of updating the same block multiple times. Maybe now we also need a `getValues` API.

ellatrix commented:
> ### 5. Changes to block bindings editor hook loop
> While this pull request, we realized that there were [some changes needed](https://github.com/WordPress/gutenberg/pull/60721/files#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R144-R169) for the block bindings editor hook. I should have created a different pull request for that. Is it okay to keep it here or is it still better to move that code to a different PR?

Let's just keep that in this PR, this is fine

ellatrix committed:
Fall back to ''

ellatrix committed:
Make sure any previous changes are persisted before resetting.

kevin940726 committed:
Add undo/redo test for resetting

kevin940726 committed:
Fix test... 🤦

ellatrix committed:
Fix remaining undo test

ellatrix committed:
Fix individual reset

SantosGuillamot commented:
Thanks a lot for chiming in ☺️ Code looks better now and the issues seem to be solved.

Regarding the new `setValues` API, I just have a couple of comments:
- I am wondering if it is worth it to keep `getValue`/`getValues` and `setValue`/`setValues` or just `getValues` and `setValues` that receive an array of all the attributes connected. It could be a bit confusing for users to decide when to use one or the other, and grouping them could be more performant and prevent unexpected issues like this one.
- I believe the [setValues function](https://github.com/WordPress/gutenberg/pull/60721/files#diff-deffea7af4722e1274cca18247b89c007e041054cc2fa315c68a0bb1425d9eb8R33) should include an array of "bindings" instead of just "attributes". Basically "bindings" could be an object including "attribute" and "args". In pattern overrides we don't have args, but we can expect it in other sources like post meta where you need to know the "key".

ellatrix commented:
@SantosGuillamot Yes, we can try that in a separate PR. Maybe one problem is that you can have args for each value for some binding sources, in which case getValue/setValue is a bit easier to use

kevin940726 committed:
Add test for undoing loses the focus

SantosGuillamot commented:
> Yes, we can try that in a separate PR. Maybe one problem is that you can have args for each value for some binding sources, in which case getValue/setValue is a bit easier to use

Yep, that's why I suggested passing a "bindings" object instead of just the attributes. This would allow us to access the "args" and let the source decide when to iterate over them.

Anyway, I'm happy to work on that in a follow-up PR if it unblocks this because the APIs are still private.

ellatrix committed:
Fix reset button test

kevin940726 commented on test/e2e/specs/editor/various/pattern-overrides.spec.js:
```diff
@@ -648,4 +702,205 @@ test.describe( 'Pattern Overrides', () => {
 		).toBeHidden();
 	} );
+
+	test( 'create a pattern with synced blocks', async ( {
```
We should probably mention "blocks with the same name" in the test title.

SantosGuillamot replied:
I've changed the description as part of [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/44b51ddb0087f2a6ef233cf2c00b0bd69cdc0cb0). Let me know if that's fine 🙂

kevin940726 commented on test/e2e/specs/editor/various/pattern-overrides.spec.js:
```diff
@@ -648,4 +702,205 @@ test.describe( 'Pattern Overrides', () => {
+		let patternId;
+		const sharedName = 'Shared Name';
+
+		await test.step( 'create a pattern with synced blocks', async () => {
```
I don't think we need this step. We already have tests for creating patterns with overrides and creating synced pattern, we don't have to repeat them here. This can be replaced with a `requestUtils` call:

```js
const innerPattern = await requestUtils.createBlock( {
	title: 'Blocks with the same name',
	content: `<!-- wp:heading {"metadata":{"name":"${ sharedName }","bindings":{"content":{"source":"core/pattern-overrides"}}}} -->
<h2>default name</h2>
<!-- /wp:heading -->
<!-- wp:paragraph {"metadata":{"name":"${ sharedName }","bindings":{"content":{"source":"core/pattern-overrides"}}}} -->
<p>default content</p>
<!-- /wp:paragraph -->
<!-- wp:paragraph {"metadata":{"name":"${ sharedName }","bindings":{"content":{"source":"core/pattern-overrides"}}}} -->
<p>default content</p>
<!-- /wp:paragraph -->`,
			status: 'publish',
		} );
```

Also, it might be interesting to see when the default contents are different? 🤔

SantosGuillamot replied:
I understood that when they share the name, we wanted to sync the values also while creating the original synced pattern, resulting in having the same default content. It'd be a bit confusing for me that the default content is not synced but whenever I use the pattern and update any of those values they become synced.

That's also why I added the two steps. The first one tests that it is synced while editing the original pattern, and the second tests that it is synced while using the pattern.

If that's not the case, we can revisit the implementation 🙂

kevin940726 replied:
I see. Could we simplify this a little bit further then?

```js
await requestUtils.createBlock( {
	title: 'Blocks with the same name',
	content: `<!-- wp:heading {"metadata":{"name":"${ sharedName }","bindings":{"content":{"source":"core/pattern-overrides"}}}} -->
<h2>default name</h2>
<!-- /wp:heading -->
<!-- wp:paragraph {"metadata":{"name":"${ sharedName }","bindings":{"content":{"source":"core/pattern-overrides"}}}} -->
<p>default content</p>
<!-- /wp:paragraph -->
<!-- wp:paragraph {"metadata":{"name":"${ sharedName }","bindings":{"content":{"source":"core/pattern-overrides"}}}} -->
<p>default content</p>
<!-- /wp:paragraph -->`,
	status: 'publish',
} );

// Update the content of one of the blocks...
await headingBlock.fill('updated content');

// Check that every content has been updated.
for ( const block of [ headingBlock, firstParagraph, secondParagraph ] ) {
	await expect( block ).toHaveContent( 'updated content' );
}
```

kevin940726 replied:
Side note: I personally still think the usage of this same name syncing feature is unknown. I don't know what the users should expect and why they would want to do this rather just creating nested patterns. 🤔

SantosGuillamot replied:
> Could we simplify this a little bit further then?

Sure, we can make that change 🙂 

> I personally still think the usage of this same name syncing feature is unknown.

I think it is a consequence of using the name as the identifier. If there are two blocks with the same name, how do you differentiate between them in the "content" object?

SantosGuillamot replied:
I've simplified the approach as suggested in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/44b51ddb0087f2a6ef233cf2c00b0bd69cdc0cb0). Let me know if that's not what you had in mind.

ellatrix commented on test/e2e/specs/editor/various/pattern-overrides.spec.js:
```diff
@@ -648,4 +702,205 @@ test.describe( 'Pattern Overrides', () => {
+
+	// A Undo/Redo bug found when implementing and fixing https://github.com/WordPress/gutenberg/pull/60721.
+	// This could be merged into an existing test after we fully test it.
+	test( 'undo/redo should not lose focuses', async ( {
```
@kevin940726 This is not really an issue related to this PR, it's also broken for other inner block instances that are controlled. For example, try editing page or content inside the site editor. It will also lose focus on undo. This seems to be an issue that we need to address in controlled inner blocks.

kevin940726 replied:
Nice find! I'm fine with removing it in this PR then 👍. Thanks!

SantosGuillamot replied:
I've removed it in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/5975f745b97b91f8fea8ae323a927b3c54c4cea5). Feel free to add it again if you change your mind 🙂

SantosGuillamot committed:
Simplify pattern synced blocks test

SantosGuillamot committed:
Remove focus test

SantosGuillamot commented:
I've been taking a deeper look, and it seems everything is working as expected, and tests are passing. Additionally, I believe the mentioned challenges seem to be solved. It'd be great to get a review to understand what is missing in order to merge this pull request.

So far, I see two follow-up tasks:
* Explore what to do with `setValue`/`setValues` API as discussed [here](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2106873747).
* Check what to do with syncDerivedUpdates function as mentioned [here](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2102508254) (point 4).

Am I missing something?

ellatrix commented:
Everything is looking good to me!

kevin940726 commented on packages/editor/src/bindings/pattern-overrides.js:
```diff
@@ -2,10 +2,82 @@
+		const parents = getBlockParents( clientId, true );
+		const patternClientId = parents.find(
+			( id ) => getBlockName( id ) === 'core/block'
+		);
```
Nit: We can just use `getBlockParentsByBlockName`.

```suggestion
		const patternClientId = getBlockParentsByBlockName( clientId, 'core/block', true )[0];
```

SantosGuillamot replied:
I changed the logic to use it in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/10fa4e8df34e2c5ba8b719227f5f0a3c9959c950). Thanks for the tip! I didn't know about that function.

kevin940726 commented on packages/editor/src/bindings/pattern-overrides.js:
```diff
@@ -2,10 +2,82 @@
+			const blockName = currentBlockAttributes?.metadata?.name;
+			const syncBlocksWithSameName = ( blocks ) => {
+				for ( const block of blocks ) {
+					if ( block.attributes?.metadata?.name === blockName ) {
```
Nit: Exit early if both names here are `undefined`.

SantosGuillamot replied:
I added an early return [here](https://github.com/WordPress/gutenberg/pull/60721/commits/3cfdd447d4328b4704a7bfeada9af205fa0a799f). I hope with that is enough. In the rest of use cases we probably want to keep iterating through the inner blocks.

kevin940726 commented on packages/editor/src/bindings/pattern-overrides.js:
```diff
@@ -2,10 +2,82 @@
+						...currentBindingValue?.[ blockName ],
+						...Object.entries( attributes ).reduce(
+							( acc, [ key, value ] ) => {
+								acc[ key ] = value === undefined ? '' : value;
```
Would be nice if we could move the comment from the original code here too.

SantosGuillamot replied:
Added in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/8c5f5748fd9b88355b74d8434e9f0764f5c88639).

kevin940726 commented on packages/patterns/src/components/reset-overrides-control.js:
```diff
@@ -6,80 +6,82 @@ import {
+			const parents = getBlockParents( props.clientId, true );
+			const patternClientId = parents.find(
+				( id ) => getBlockName( id ) === 'core/block'
+			);
```
Nit: Same here.

```suggestion
			const patternClientId = getBlockParentsByBlockName( props.clientId, 'core/block', true );
```

SantosGuillamot replied:
Done as part of [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/10fa4e8df34e2c5ba8b719227f5f0a3c9959c950).

kevin940726 commented on test/e2e/specs/editor/various/pattern-overrides.spec.js:
```diff
@@ -148,6 +148,8 @@ test.describe( 'Pattern Overrides', () => {
 				name: 'Block: Paragraph',
 			} );
+			// Ensure the first pattern is selected.
+			await patternBlocks.first().selectText();
```
Why do we need this?

SantosGuillamot replied:
It seems that if I don't select the pattern, the blocks always have `inert="true"`.

kevin940726 replied:
Hmm, interesting! I guess I just didn't notice it before 😅.

kevin940726 commented on test/e2e/specs/editor/various/pattern-overrides.spec.js:
```diff
@@ -429,41 +433,22 @@ test.describe( 'Pattern Overrides', () => {
+		// Check it renders correctly.
+		const headingBlock = editor.canvas.getByRole( 'document', {
+			name: 'Block: Heading',
+			includeHidden: true,
```
Is this needed?

SantosGuillamot replied:
I don't think so. Removed in [this commit](https://github.com/WordPress/gutenberg/pull/60721/commits/33bfb866f3d7995bab5218f72783bf3d0262b003).

kevin940726 commented on test/e2e/specs/editor/various/patterns.spec.js:
```diff
@@ -254,22 +279,15 @@ test.describe( 'Synced pattern', () => {
-			},
-		] );
+		const [ syncedPattern ] = await editor.getBlocks();
+		const patternRecord = await getPatternRecord(
```
Nit: Same here. I don't think we need to retest this. We can just remove this.

SantosGuillamot replied:
Sorry, I totally missed [the previous comment](https://github.com/WordPress/gutenberg/pull/60721#discussion_r1584042125). I've modified the logic to match what you suggested (I hope 😄 ).
- Checking that the "core/block" is present and check the frontend in the first test: [link](https://github.com/WordPress/gutenberg/pull/60721/commits/d80aba74b6f9f4eb55cc0209e4e8f55e9850b72b).
- Checking the frontend also when we modify the pattern: [link](https://github.com/WordPress/gutenberg/pull/60721/commits/6b97321200cf7f0f7c225e2738fa2c1d309c4b8f).
- Just check that the "core/block" is present in the rest of the tests: [link](https://github.com/WordPress/gutenberg/pull/60721/commits/c122b7991898c00b8d85fa59d174f3950aaca1cf).

Let me know if that's not what you had in mind 🙂

kevin940726 replied:
Yes! Thank you for improving the tests ❤️!

We could also check the editor but I don't think it's necessary. It's basically just calling `getBlocks()` and we'd just be testing implementation details. I think the tests now should cover almost everything.

kevin940726 commented on test/e2e/specs/editor/various/patterns.spec.js:
```diff
@@ -359,19 +381,24 @@ test.describe( 'Synced pattern', () => {
 			attributes: { ref: id },
 		} );
 
+		// // Wait until the pattern is created and inserted.
```
Nit:
```suggestion
		// Wait until the pattern is created and inserted.
```

kevin940726 reviewed (approved):
I think this looks good and we can follow-up with other PRs 💯!

If we encounter any other issues we can also add e2e tests for them.

kevin940726 commented:
> Check what to do with syncDerivedUpdates function as mentioned [here](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2102508254) (point 4).

Oh yeah, `syncDerivedUpdates` is only created to solve this issue and not being used anywhere else for now. It's safe to delete it if it's not needed anymore.

SantosGuillamot committed:
Use getBlockParentsByBlockName

SantosGuillamot committed:
Add early return if blockName doesn't exist

SantosGuillamot committed:
Add comment for undefined values

SantosGuillamot committed:
Remove comment typo

SantosGuillamot committed:
Remove unnecessary includeHidden from tests

SantosGuillamot committed:
Check the frontend in patterns test

SantosGuillamot committed:
Check frontend in updated pattern

SantosGuillamot committed:
Stop using `getPatternRecord`

SantosGuillamot reviewed (approved):
I believe I've addressed the remaining comments. If the tests pass, I would say this is ready to be merged.

ellatrix commented on packages/block-library/src/block/edit.js:
```diff
@@ -401,47 +215,16 @@ function ReusableBlockEdit( {
+		value: innerBlocks.length > 0 ? innerBlocks : blocks,
+		onInput: NOOP,
+		onChange: NOOP,
+		renderAppender: blocks?.length ? undefined : blocks.ButtonBlockAppender,
```
Is this a typo?

ellatrix replied:
(`blocks.ButtonBlockAppender` will never exist)

SantosGuillamot replied:
It seems that change has been there since the [first commit](https://github.com/WordPress/gutenberg/pull/60721/commits/78f960185df3dc1c3d26f084f9375b4431444fab), so I assume we just replaced innerBlocks by blocks. Maybe we should go back to the previous logic?

```
renderAppender: innerBlocks?.length
    ? undefined
    : InnerBlocks.ButtonBlockAppender,
```

ellatrix commented on test/e2e/specs/editor/various/patterns.spec.js:
```diff
@@ -191,7 +179,29 @@ test.describe( 'Synced pattern', () => {
+
+		// Check that the frontend shows the content of the pattern.
+		const postId = await editor.publishPost();
+		await page.goto( `/?p=${ postId }` );
```
Is there no preview util? This won't work with subdirectory installations, and breaks running tests in MAMP. We should honour the WP_BASE_URL.

SantosGuillamot replied:
It seems there is a [openPreviewPage util](https://github.com/WordPress/gutenberg/blob/800c96b18a9e18162d1fbd29fb2a0cd51f950778/packages/e2e-test-utils-playwright/src/editor/preview.ts#L18). I can try to use that in a follow-up PR.

ellatrix commented:
Let's just create follow-up PRs

gziolo commented:
So great to see this refactoring landed. Excellent collaboration! 🎉 

> Explore what to do with setValue/setValues API as discussed https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2106873747.

One important aspect to keep in mind is that the current implementation of `setValues` can be applicable only for Pattern Overrides sources that don't use any custom arguments (`args`). For most of the other popular sources, it's going to be the opposite as these additional params inform the block bindings source handler, which custom field to update in the database on the server. Eventually, the final public API should contain only one of the following pairs: `getValues`/`setValues`, `getValue`/`setValues` or `getValue`/`setValue`.

SantosGuillamot commented:
> Eventually, the final public API should contain only one of the following pairs: getValues/setValues, getValue/setValues or getValue/setValue.

I agree. I plan to work on a follow-up PR for that, and we can discuss the best approach there. My idea is to change the current `setValues` implementation to receive the whole binding (including args) and not just the attributes.

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -401,47 +215,16 @@ function ReusableBlockEdit( {
-		renderAppender: innerBlocks?.length
-			? undefined
-			: InnerBlocks.ButtonBlockAppender,
+		value: innerBlocks.length > 0 ? innerBlocks : blocks,
```
@ellatrix, @SantosGuillamot, can you provide additional context around this change?

> Use `blocks` variable until `innerBlocks` is populated, which has the proper clientIds.

What does proper clientIds mean here?

This seems to be the cause of the infinite loop regression in #63909.

talldan replied:
It's possibly related to the way `useBlockSync` clones the inner blocks for controlled blocks:
https://github.com/WordPress/gutenberg/blob/54eee32b043ff1879cba2f82f9d5568df43bc34c/packages/block-editor/src/components/provider/use-block-sync.js#L114

That gives the blocks new client ids.

Having new client ids is important. If you insert two instances of the same pattern into a post the inner blocks need unique client ids so that updates to one don't update the other.

SantosGuillamot replied:
Sorry, I've been away for a few days 🙂 

I remember having issues with having different `clientIds` in `blocks` and in `innerBlocks` as explained in [this comment](https://github.com/WordPress/gutenberg/pull/60721#issuecomment-2077765846). I even refer to these lines there. It seems the `innerBlocks` clientIds were actually the ones in the markup, that's what I was trying to refer with "proper" clientIds.

What I don't remember is if I added that check because it was breaking something or just because I thought it was correct.

I'm fine trying to remove it and see if the tests pass.
